Update tests to match current alpha node-logs feature

fabi200123 opened the pull request:
What type of PR is this?
/kind feature

What this PR does / why we need it:
This PR modifies the e2e tests for the feature: node log viewer to match the current format of the feature.
The --service and --path were collapsed into --query
It also adds tests for the --role and --raw options

Ref:
Feature PR Link: https://github.com/kubernetes/kubernetes/pull/96120

Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:
KEP: https://github.com/kubernetes/enhancements/pull/2271
KEP Discussion: https://github.com/kubernetes/enhancements/issues/2258
Usage: https://github.com/kubernetes/website/pull/32660

fabi200123 committed:
Updated and added more tests

Updated the tests and added more test for the Node Log Viewer feature

aravindhp commented:
/ok-to-test
/sig-windows

marosset commented:
/test ?

SergeyKanzhelev commented:
/assign @fromanirh 

/priority important-longterm
/triage accepted

haircommander commented on test/e2e/node/kubelet.go:
```diff
@@ -479,62 +484,215 @@ var _ = SIGDescribe("kubelet", func() {
 		/*
-			Test if kubectl node-logs <node-name> --path pods
-			returns something or not!
+			Test if kubectl node-logs --role master --query docker --tail 4
+			returns 4 lines from the docker logs or not!
```
hm maybe we should do `containerd` or `crio` depending on which the node is running now that dockershim support is dropped

haircommander replied:
or rather, maybe we should just use `kubelet` as that's guaranteed to be running, and the test isn't testing whether we can get specific logs from the runtime, but that we get the tail of the logs in question

fabi200123 replied:
Changed it to kubelet. Thank you for pointing this out!

haircommander commented on test/e2e/node/kubelet.go:
```diff
@@ -479,62 +484,215 @@ var _ = SIGDescribe("kubelet", func() {
+			result := runKubectlCommand(cmd, "--role --query --tail --output")
+			framework.ExpectEqual(len(lines(result)), 1)
+			lookForString("docker", result)
+			r := checkJSON(result)
+			if r != nil {
+				framework.Failf("Receive invalid json: %v", r)
+			}
```
nit: can we compress this to `if r := checkJSON...; r != nil {`

haircommander commented:
/retest

I have a couple of nits. throughout: it looks like these tests are testing cli stability and not always behavior. It may be good to consider ways we can test node-logs is retrieving what we're asking for (whenever possible)

haircommander commented:
you even could start a dummy service on the node with a systemd transient unit, and get the node logs from that. That way, we can test things like `hostname` (for the `-l` test) or the node role

fabi200123 commented on test/e2e/node/kubelet.go:
```diff
@@ -479,62 +485,267 @@ var _ = SIGDescribe("kubelet", func() {
+			cmd := tk.KubectlCmd("alpha", "node-logs", "--role", "master", "--query", "kubelet", "--tail", n, "-o", "short")
+			result := runKubectlCommand(cmd, "--role --query --tail")
+			logs := journalctlCommand("-u", "kubelet", "-n", n, "")
+			framework.Logf("LOG LOG LOG")
```
Yeah, sorry for that. Thought I deleted it

haircommander commented:
still
/lgtm

SergeyKanzhelev commented on test/e2e/node/kubelet.go:
```diff
@@ -479,62 +485,266 @@ var _ = SIGDescribe("kubelet", func() {
+			tk := e2ekubectl.NewTestKubeconfig(framework.TestContext.CertDir, framework.TestContext.Host, framework.TestContext.KubeConfig, framework.TestContext.KubeContext, framework.TestContext.KubectlPath, ns)
+			cmd := tk.KubectlCmd("alpha", "node-logs", "--role", "master", "--query", "kubelet", "--tail", "1", "--output", "json")
+			result := runKubectlCommand(cmd, "--role --query --tail --output")
+			framework.ExpectEqual(len(lines(result)), 1)
```
this is the only use of `lines`. Does it have to be defined as a func 100 lines above?

SergeyKanzhelev commented on test/e2e/node/kubelet.go:
```diff
@@ -466,6 +469,9 @@ var _ = SIGDescribe("kubelet", func() {
 			numNodes  int
 			nodeNames sets.String
 		)
+		lines := func(out string) []string {
```
not related to this PR, but can you please update a feature tag on the test to something without spaces. Like `[Feature:NodeLogViewer]` to match the style of other names. Extra bonus for using the feature gate name with this tag

SergeyKanzhelev commented on test/e2e/node/kubelet.go:
```diff
@@ -479,62 +485,266 @@ var _ = SIGDescribe("kubelet", func() {
+				}
+				err = cmd.Wait()
+				if err == nil {
+					framework.Failf("Command failed: %v. It was expected an error!", cmd)
```
I wonder what `%v` will output for `*exec.Cmd` type. Not sure it will be readable. Can you show the log example?

fabi200123 replied:
The output will be the command ran: `Command failed: /usr/bin/kubectl --server=... --kubeconfig=/etc/kubernetes/admin.conf --namespace=kubelet-9887 alpha node-logs <node-name>

SergeyKanzhelev commented on test/e2e/node/kubelet.go:
```diff
@@ -479,62 +485,266 @@ var _ = SIGDescribe("kubelet", func() {
+			result := runKubectlCommand(cmd, "--role --query --tail")
+			logs := journalctlCommand("-u", "kubelet", "-n", n, "")
+			if result != logs {
+				framework.Failf("Failed to receive the correct kubelet logs or the correct amount of lines of logs")
```
would be useful to output both compared values to simplify test troubleshooting

fabi200123 replied:
`runKubectlCommand` already logs the output. I will modify the message and add a framework.Logf to the journalctl function as well

SergeyKanzhelev commented on test/e2e/node/kubelet.go:
```diff
@@ -479,62 +485,266 @@ var _ = SIGDescribe("kubelet", func() {
+			result := runKubectlCommand(cmd, "--role --query --pattern")
+			logs := journalctlCommand("-u", "kubelet", "-b", "0", "-g", "kubelet", "")
+			if result != logs {
+				framework.Failf("Failed to receive the correct kubelet logs for the specific pattern")
```
also would be nice to have compared values as part of a test failure message or in logs

SergeyKanzhelev commented on test/e2e/node/kubelet.go:
```diff
@@ -479,62 +485,266 @@ var _ = SIGDescribe("kubelet", func() {
+			returns the kubelet logs or not!
+		*/
+
+		ginkgo.It("should return the logs for the requested query using the selector kubernetes.io/os=linux", func() {
```
Please apply `[LinuxOnly]` tag

SergeyKanzhelev commented on test/e2e/node/kubelet.go:
```diff
@@ -479,62 +485,266 @@ var _ = SIGDescribe("kubelet", func() {
+			tk := e2ekubectl.NewTestKubeconfig(framework.TestContext.CertDir, framework.TestContext.Host, framework.TestContext.KubeConfig, framework.TestContext.KubeContext, framework.TestContext.KubectlPath, ns)
+			cmd := tk.KubectlCmd("alpha", "node-logs", "--role", "master", "--query", "kubelet", "-o", "short", "--until-time", start.Format(RFC3339))
+			result := runKubectlCommand(cmd, "--role --query --until-time")
+			logs := journalctlCommand("-u", "kubelet", "-U", "now", "")
```
if this test is run on soak test environment, there will be a LOT of logs. The choice of date to be 2006 may not be ideal.

fabi200123 replied:
This will print the logs until the current time of the test. By default, kubectl alpha node-logs prints the logs of the current boot. The 2006 is the default date to declare the RFC3339 from what I have read. And if you run the command for --since-time or --until-time with an invalid date, the output is:
`kubectl alpha node-logs -Q query --since-time 20
error: parsing time "20" as "2006-01-02T15:04:05Z07:00": cannot parse "20" as "2006"`

SergeyKanzhelev commented on test/e2e/node/kubelet.go:
```diff
@@ -479,62 +485,266 @@ var _ = SIGDescribe("kubelet", func() {
 	stdout, stderr, err := framework.StartCmdAndStreamOutput(cmd)
+	var buf bytes.Buffer
 	if err != nil {
 		framework.Failf("Failed to start kubectl command: %v", err)
```
would it be interesting to dump stderr in log when command failed?

SergeyKanzhelev commented on test/e2e/node/kubelet.go:
```diff
@@ -479,62 +485,266 @@ var _ = SIGDescribe("kubelet", func() {
+
+	b_read, err := io.Copy(&buf, stdout)
+	if err != nil {
 		framework.Failf("Expected output from kubectl alpha node-logs %s: %v", arg, err)
```
would it be interesting to dump stderr in log when there is no stdout?

SergeyKanzhelev commented on test/e2e/node/kubelet.go:
```diff
@@ -479,62 +485,266 @@ var _ = SIGDescribe("kubelet", func() {
+	if strings.Contains(result, expectedString) {
+		return
+	}
+	framework.Failf("Failed to find \"%s\", last result: \"%s\"", expectedString, result)
```
what does "last result" means in this context? Did you mean:

```suggestion
	framework.Failf("Failed to find \"%s\" in \"%s\"", expectedString, result)
```

SergeyKanzhelev commented on test/e2e/node/kubelet.go:
```diff
@@ -479,62 +485,266 @@ var _ = SIGDescribe("kubelet", func() {
-				runKubectlCommand(cmd, "--service")
+				cmd := tk.KubectlCmd("alpha", "node-logs", nodeName, "--query", "kubelet")
+				result := runKubectlCommand(cmd, "--query")
+				lookForString("kubelet", result)
```
may need a better name than `lookForString`. Maybe `assertContains`

SergeyKanzhelev commented on test/e2e/node/kubelet.go:
```diff
@@ -479,62 +485,266 @@ var _ = SIGDescribe("kubelet", func() {
+	if strings.Contains(result, expectedString) {
+		return
+	}
+	framework.Failf("Failed to find \"%s\", last result: \"%s\"", expectedString, result)
```
if `result` is a megabyte of logs, will test framework be ok with it? How would the failure message look like in this case? Should the result be logged instead of being part of the test failure message?

fabi200123 replied:
Yes, the command `runKubectlCommand` already logs the result. I will modify the function to output: **"Failed to find the expectedString"**

SergeyKanzhelev reviewed (commented):
functionality is generally ok. I have styling and test debuggability comments

SergeyKanzhelev commented:
/lgtm remove

removing lgtm to avoid placing this into Approver Needed column on triage

fabi200123 committed:
Modify tests to match UTC timezone

The kubectl alpha node-logs return the logs in UTC timezone, so the tests should be modified to match that timezone. (Journalctl returns the logs in local timezone)

dims commented:
If you still need this PR then please rebase, if not, please close the PR

SergeyKanzhelev commented:
@fabi200123 this is a nice PR. Do you plan to rebase and make it ready for merge?

aravindhp commented:
@SergeyKanzhelev the plan is to pull this into https://github.com/kubernetes/kubernetes/pull/96120/ but I am trying to figure out the best way to move forward there given this [comment](https://github.com/kubernetes/kubernetes/pull/96120#pullrequestreview-1262371094). 

@fabi200123 I suggest we close this for now. We can figure out a way to cherry-pick the test based on the outcome of the feature.
